dot/network/notifications.go

```diff
@@ -289,16 +289,15 @@ func (s *Service) sendData(peer peer.ID, hs Handshake, info *notificationsProtoc
 			return
 		}
 
-		// TODO: ensure grandpa stores *all* previously received votes and discards them
-		// only when they are for already finalised rounds; currently this causes issues
-		// because a vote might be received slightly too early, causing a round mismatch err,
-		// causing grandpa to discard the vote. (#1855)
-		_, isConsensusMsg := msg.(*ConsensusMessage)
-		if !added && !isConsensusMsg {
+		// if we could not add message in cache, message was already present in the
+		// cache and was sent.
+		if !added {
 			return
 		}
 	}
 
+	// TODO: Add message to cache after we successfully wrote it to stream?
```

Comment:
the message was already written to the cache in line 286

Reply:
oh this is a good point actually - we should only write it to the cache if it was sent successfully. can you put a `has` check at line 286 instead, and move the `put` after the `writeToStream`?

Reply:
addressed